mmseg/models/decode_heads/ham_head.py

```diff
@@ -0,0 +1,245 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+# Originally from https://github.com/visual-attention-network/segnext
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from mmcv.cnn import ConvModule
+
+from mmseg.ops import resize
+from ..builder import HEADS
+from .decode_head import BaseDecodeHead
+
+
+class _MatrixDecomposition2DBase(nn.Module):
+
+    def __init__(self,
+                 spatial=True,
+                 MD_S=1,
+                 MD_D=512,
+                 MD_R=64,
+                 train_steps=6,
+                 eval_steps=7,
+                 inv_t=100,
+                 eta=0.9,
+                 rand_init=True):
+        super().__init__()
+
+        self.spatial = spatial
+
+        self.S = MD_S
+        self.D = MD_D
+        self.R = MD_R
+
+        self.train_steps = train_steps
+        self.eval_steps = eval_steps
+
+        self.inv_t = inv_t
+        self.eta = eta
+
+        self.rand_init = rand_init
+
+        print('spatial', self.spatial)
+        print('S', self.S)
+        print('D', self.D)
+        print('R', self.R)
+        print('train_steps', self.train_steps)
+        print('eval_steps', self.eval_steps)
+        print('inv_t', self.inv_t)
+        print('eta', self.eta)
+        print('rand_init', self.rand_init)
+
+    def _build_bases(self, B, S, D, R, cuda=False):
+        raise NotImplementedError
+
+    def local_step(self, x, bases, coef):
+        raise NotImplementedError
+
+    # @torch.no_grad()
+    def local_inference(self, x, bases):
+        # (B * S, D, N)^T @ (B * S, D, R) -> (B * S, N, R)
+        coef = torch.bmm(x.transpose(1, 2), bases)
+        coef = F.softmax(self.inv_t * coef, dim=-1)
+
+        steps = self.train_steps if self.training else self.eval_steps
+        for _ in range(steps):
+            bases, coef = self.local_step(x, bases, coef)
+
+        return bases, coef
+
+    def compute_coef(self, x, bases, coef):
+        raise NotImplementedError
+
+    def forward(self, x, return_bases=False):
+        B, C, H, W = x.shape
+
+        # (B, C, H, W) -> (B * S, D, N)
+        if self.spatial:
+            D = C // self.S
+            N = H * W
+            x = x.view(B * self.S, D, N)
+        else:
+            D = H * W
+            N = C // self.S
+            x = x.view(B * self.S, N, D).transpose(1, 2)
+
+        cuda = x.device == torch.device('cuda')
```

Comment:
This assertion may be raise `cuda=False` **in multiple GPUs case** where `device(type='cuda', index=x)` != `device(type='cuda')`.
![image](https://user-images.githubusercontent.com/24582831/205248959-fbb3fc10-5194-4d52-8d36-6e52c9241b29.png)

By the way, in your 2 GPUs && SyncBN environment, did this line not raise error?

![image](https://user-images.githubusercontent.com/24582831/205249499-c16814cf-0eb7-403e-9986-2c4c711e21da.png)